### ISSUE TITLE
potential data integrity issue if storage options are customized during role execution

### ISSUE THREAD

**Issue by jonseymour** [angstwad/docker.ubuntu#89]
I think have found an issue if this role is used together with custom storage options designed to enable direct-lvm mode  https://docs.docker.com/engine/userguide/storagedriver/device-mapper-driver/

The issue arises because after the DOCKER_OPTS variable is configured, there is no restart of the docker daemon. As a result, the initial install of the docker daemon will be running with the thin pool device mapped to the loop back device instead of the thin pool device of the intended volume group. When the docker daemon is subsequently restarted the thin pool device implied by the storage options will be used which, of course, is a different device.

**Comment by jonseymour** [comment]
For example, here is a case where lsblk shows that the docker pool device is mapped across loopback devices when it should be mapped across logical volumes from the datavg volume group

```
NAME                            MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda                               8:0    0  29.3G  0 disk 
└─sda1                            8:1    0  29.3G  0 part /
sdb                               8:16   0    14G  0 disk 
└─sdb1                            8:17   0    14G  0 part /mnt
sdc                               8:32   0   512G  0 disk 
├─datavg-spherestack (dm-0)     252:0    0   128G  0 lvm  /var/lib/sphere-stack
├─datavg-backups (dm-1)         252:1    0   128G  0 lvm  /var/lib/backups
├─datavg-docker--data (dm-2)    252:2    0   128G  0 lvm  
└─datavg-docker--meta (dm-3)    252:3    0     4G  0 lvm  
sdd                               8:48   0   128G  0 disk 
└─sdd1                            8:49   0   128G  0 part 
sr0                              11:0    1   1.1M  0 rom  
loop0                             7:0    0   100G  0 loop 
└─docker-8:1-528619-pool (dm-4) 252:4    0   100G  0 dm   
loop1                             7:1    0     2G  0 loop 
└─docker-8:1-528619-pool (dm-4) 252:4    0   100G  0 dm   
```

**Comment by jonseymour** [comment]
I did some more digging. The issue is not so much that the docker daemon is restarted - it is. The problem is that the thin pool device is not torn down so that when docker restarts with the new storage options, the thin pool device is still bound to the loop back devices.

So:

```
$ sudo dmsetup table
datavg-spherestack: 0 268435456 linear 8:32 276826112
datavg-docker--data: 0 268435456 linear 8:32 2048
docker-8:1-528625-pool: 0 209715200 thin-pool 7:1 7:0 128 32768 1 skip_block_zeroing 
datavg-docker--meta: 0 8388608 linear 8:32 268437504
datavg-backups: 0 268435456 linear 8:32 545261568
$ lsblk
NAME                            MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT
sda                               8:0    0  29.3G  0 disk 
└─sda1                            8:1    0  29.3G  0 part /
sdb                               8:16   0    14G  0 disk 
└─sdb1                            8:17   0    14G  0 part /mnt
sdc                               8:32   0   512G  0 disk 
├─datavg-spherestack (dm-0)     252:0    0   128G  0 lvm  /var/lib/sphere-stack
├─datavg-backups (dm-1)         252:1    0   128G  0 lvm  /var/lib/backups
├─datavg-docker--data (dm-2)    252:2    0   128G  0 lvm  
└─datavg-docker--meta (dm-3)    252:3    0     4G  0 lvm  
sdd                               8:48   0   128G  0 disk 
└─sdd1                            8:49   0   128G  0 part 
sr0                              11:0    1   1.1M  0 rom  
loop0                             7:0    0   100G  0 loop 
└─docker-8:1-528625-pool (dm-4) 252:4    0   100G  0 dm   
loop1                             7:1    0     2G  0 loop 
└─docker-8:1-528625-pool (dm-4) 252:4    0   100G  0 dm   

```

**Comment by jonseymour** [comment]
Ok, the issue in  my case is that my host is ubuntu 14.04 and these lines:

```
- name: Set docker daemon options
  copy:
    content: "DOCKER_OPTS=\"{{ docker_opts.rstrip('\n') }}\""
    dest: /etc/default/docker
    owner: root
    group: root
    mode: 0644
  notify:
    - Reload docker
  when: docker_opts != "" and ansible_distribution_version|version_compare(15.04, '<')


```

don't actually result in the restart of the docker daemon with the new storage options. That is: "Reload docker" silently fails. "Restart docker" would likely have the desired effect, which I will now test.

**Comment by angstwad** [comment]
Closed by #90
